### PR TITLE
pkg/archive: remove backward compat hack for go < 1.9

### PR DESCRIPTION
**Pull request by thaJeztah**
- reverts https://github.com/moby/moby/pull/33935
- reverts https://github.com/moby/moby/pull/32959
- relates to https://github.com/moby/moby/pull/11148
- relates to https://github.com/moby/buildkit/pull/152

The fillGo18FileTypeBits func was added in 1a451d9a7bb9cd7d437b42d4e73b0560fbf84348 (https://github.com/moby/moby/pull/33935)
to keep the tar headers consistent with headers created with go1.8 and older.

go1.8 and older incorrectly preserved all file-mode bits, including file-type,
instead of stripping those bits and only preserving the _permission_ bits, as
defined in;

- the GNU tar spec: https://www.gnu.org/software/tar/manual/html_node/Standard.html
- and POSIX: http://pubs.opengroup.org/onlinepubs/009695399/basedefs/tar.h.html

We decided at the time to copy the "wrong" behavior to prevent a cache-bust and
to keep the archives identical, however:

- It's not matching the standards, which causes differences between our tar
  implementation and the standard tar implementations, as well as implementations
  in other languages, such as Python (see docker/compose#883).
- BuildKit does not implement this hack.
- We don't _need_ this extra information (as it's already preserved in the
  type header; https://pkg.go.dev/archive/tar#pkg-constants

In short; let's remove this hack.

This reverts commit 1a451d9a7bb9cd7d437b42d4e73b0560fbf84348.
This reverts commit 41eb61d5c2a44c745b11ada948034fd3bcc42db5.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
I guess this one was expected (will have a look)

```

[2022-09-01T23:59:51.673Z] === Failed
[2022-09-01T23:59:51.673Z] === FAIL: builder/remotecontext TestHashFile (0.02s)
[2022-09-01T23:59:51.673Z]     tarsum_test.go:66: invalid checksum. expected 1149ab94af7be6cc1da1335e398f24ee1cf4926b720044d229969dfc248ae7ec, got 55dfeb344351ab27f59aa60ebb0ed12025a2f2f4677bf77d26ea7a671274a9ca
[2022-09-01T23:59:51.673Z] 
[2022-09-01T23:59:51.673Z] === FAIL: builder/remotecontext TestHashSubdir (0.02s)
[2022-09-01T23:59:51.673Z]     tarsum_test.go:103: invalid checksum. expected d7f8d6353dee4816f9134f4156bf6a9d470fdadfb5d89213721f7e86744a4e69, got 74a3326b8e766ce63a8e5232f22e9dd895be647fb3ca7d337e5e0a9b3da8ef28
[2022-09-01T23:59:51.673Z] 
```

File: pkg/archive/archive_windows.go [hunk 1]
```diff
@@ -31,14 +31,11 @@ func CanonicalTarNameForPath(p string) string {
...
-	noPermPart := perm &^ os.ModePerm
-	// Add the x bit: make everything +x from windows
-	permPart |= 0111
-	permPart &= 0755
+	// Remove group- and world-writable bits.
+	perm &= 0o755
 
-	return noPermPart | permPart
+	// Add the x bit: make everything +x on Windows
+	return perm | 0o111
```
Comment (thaJeztah) [inline pkg/archive/archive_windows.go:38]: 1. Commit 1a22418f9f1573ca8521831d52c8e0562cb3ef8f (https://github.com/moby/moby/pull/11148) changed permissions to `0700`
   on Windows, or more factually, it removed `rw` (`chmod g-rw,o-rw`) and added
   executable bits (`chmod u+x`) to address https://github.com/moby/moby/issues/11047.
2. This was too restrictive, and b7dc9040f04fe8dacd2f14c6b93d1b0bb6bde333 (https://github.com/moby/moby/pull/11395) changed
   permissions to only remove the group- and world-writable bits to give read and
  execute access to everyone, but setting execute permissions for everyone to address https://github.com/moby/moby/issues/11246
3. However, this also removed the non-permission bits, so 41eb61d5c2a44c745b11ada948034fd3bcc42db5 (https://github.com/moby/moby/pull/32959)
   updated the code to preserve those, and keep parity with Linux.

This changes it back to `2.`. I wonder (_think_) _permission_ bits (read, write)
can be portable, except for the _executable_ bit (which is not present on Windows).
The alternative could be to keep the permission bits, and only set the executable
bit (`perm | 0111`) for everyone (equivalent of `chmod +x`), but that likely would
be a breaking change.

**Comment by thaJeztah** [comment]
@tonistiigi @corhere ptal (this was the change we were discussing in the maintainers call)

**Review by corhere — APPROVED** [review]

**Review by cpuguy83 — APPROVED** [review]
LGTM, but maybe make sure we get some extra eyes on this.

@tianon @tonistiigi do you see any issue here?

**Review by tonistiigi — APPROVED** [review]
